### ISSUE TITLE
[transformers] model integration

### ISSUE THREAD

**Issue by eustlb** [facebookresearch/omnilingual-asr#5]
### Feature Description

Hey,
What do you think about raising a PR to integrate the model into [transformers](https://github.com/huggingface/transformers)?

This way they could be used simply by simply pip installing `transformers` and:
```python
from transformers import OmnilingualASR

model = OmnilingualASR.from_pretrained("facebook/omnilingual_asr")
model.generate(...)
```

It should be fairly easy since we already have models with similar architectures, and it would make the models easier to use and more accessible. Happy to guide you through the process!

**Comment by artemru** [comment]
Hello, @eustlb ! No doubt - this integration will be super useful !
We considered it at some point but it turned out to be incompatible with release timeline and priorities. 
So, thanks for offering your assistance ) we'll be happy to get some help for the integration process ! what is the best way to connect in order to plan that work ?

**Comment by StephennFernandes** [comment]
@artemru @eustlb 

hey guys just wanted to know on what makes porting the following model to HF Transformers so difficult ? is the GLUE function or the ASR LLM joint mapping ?

**Comment by eustlb** [comment]
Hey @StephennFernandes, nothing complicated, just a bandwidth matter... 😁

**Comment by StephennFernandes** [comment]
Hey @eustlb i totally understand, although i haven't ever contributed a model to HF, if you could guide me, i would love to take up the task and contribute these models to HF . 

i promise to put if significant amount of efforts and hold up to  high standards of code quality as you expect. 
would overall love to have this opportunity, given some guidance.

**Comment by harikc456** [comment]
@eustlb I've been working on porting the model to Hugging Face, with my scripts available in [this repo](https://github.com/harikc456/wav2vec2_llama_hf). However, I'm unable to reproduce the same transcriptions as the fairseq model.

During debugging, I noticed a discrepancy in audio normalization between the two implementations: `Wav2Vec2FeatureExtractor` uses `zero_mean_unit_var_norm`, while `omnilingual-asr` uses PyTorch's `layer_norm`. This causes the inputs to differ slightly between the models.

Has anyone else encountered this issue, or is there a reason for this difference in normalization approaches?

**Comment by ebezzam** [comment]
@harikc456 thanks for working on porting the LLM-decoder variant! I'm from HF, working with @eustlb on audio model integrations into Transformers.

I've opened a PR where I've been on working on the CTC-variant integration: https://github.com/huggingface/transformers/pull/43265
(we'd likely also add the LLM-variant to this PR)

_Note it is very scrappy at the moment, as I'm figure out the best design._ In short, when adding new models to Transformers, we try to use a tool called [modular](https://huggingface.co/docs/transformers/en/modular_transformers#contributing-a-new-model-to-transformers) to build off existing models and preventing code duplication. However, Wav2Vec2 was added to Transformers before `fairseq2` was introduced. So I'm taking this opportunity for refactoring to make the model integration more compatible/intuitive with `fairseq2`. Nevertheless, what I'm doing in the PR will be useful for also integrating the LLM-decoder variant (as you've started 🙏 ).

Re discrepancy in audio normalization: I've noticed the same thing. In this case, we can define a new feature extractor, see [here](https://github.com/ebezzam/transformers/blob/4de43c70e1eb656e9c206a9701d85d7a61e79486/src/transformers/models/omniasr/feature_extraction_omniasr.py#L238) where I use `layer_norm` in a new `OmniASRFeatureExtractor`. For CTC-decoding, this normalization difference didn't lead to different outputs. _What did lead to different outputs for me was differences in the model dtype_. By default, Omnilingual uses `bfloat16` (see [here](https://github.com/facebookresearch/omnilingual-asr/blob/81f51e224ce9e74b02cc2a3eaf21b2d91d743455/src/omnilingual_asr/models/inference/pipeline.py#L157)), whereas I notice [your checkpoint](https://huggingface.co/harikc456/wav2vec2-llama-300m) is saved (and probably loaded?) as float32, which could make a difference. (But it could also be that the LLM-variant is more sensitive to the input difference and the output difference will still be there, haven't tried yet). In any case, it's good to make sure the dtype of the original and the Transformers model are the same when comparing!

**Comment by harikc456** [comment]
@ebezzam Thanks for sharing your draft implementation of the CTC variant models. I reused the feature extractor from your branch, which fixed the input mismatch issue.

I've made several other architecture-related fixes, but I'm still unable to match the output of the fairseq2 checkpoint. Through analysis, I've pinpointed that the divergence occurs at the `self_attn` layer of llama decoder. Specifically, the Q, K, V inputs differ from fairseq2's values, even though the outputs from `q_proj`, `k_proj`, and `v_proj` layers are identical. This suggests the RoPE encodings might be the source of divergence.

Looking at the fairseq2 implementation, I found two potential differences:
1. fairseq2 uses the complex number variant of RoPE (which should be mathematically equivalent)
2. fairseq2 adds +1 to `position_indices` for padding

Have you encountered similar issues when porting other models or LLMs? Any insights would be appreciated.

**Comment by harikc456** [comment]
@ebezzam I've successfully achieved transcription parity with the fairseq2 implementation! 

After thorough testing with both the 300M LLM and CTC models, I can confirm that the transcriptions are consistent across implementations. Note that while the final transcriptions match, there are slight differences in the logits between the two implementations.

**Models:**
- [Omnilingual CTC 300M v2](https://huggingface.co/harikc456/omnilingual-ctc-300m-v2)
- [Omnilingual LLM 300M v2](https://huggingface.co/harikc456/wav2vec2-llama-300m)

**Architecture:**
- [Implementation details](https://github.com/harikc456/wav2vec2_llama_hf/tree/main)

**Comment by ebezzam** [comment]
@harikc456 great! From my PR, I was able to get consistent results with the CTC variant. I was in the middle of doing the LLM variant, but had to switch to work on another model. I'll try to take a look later this week / next week!

**Comment by StephennFernandes** [comment]
@ebezzam any updates on this ?

**Comment by ebezzam** [comment]
@StephennFernandes some other projects/priorities came up 😓 it will probably be a couple weeks till I can get back to it. Sorry for the delay!

**Comment by StephennFernandes** [comment]
@ebezzam no worries !!!

**Comment by StephennFernandes** [comment]
@ebezzam hey any luck on progressing through this ?

**Comment by StephennFernandes** [comment]
hey @ebezzam sorry for being pushy but any update on this ? when can we expect the v2 latest variant on the hub ?

**Comment by ebezzam** [comment]
@StephennFernandes thanks for checking in, let's shift the conversation to the ongoing PR: https://github.com/huggingface/transformers/pull/43265

We haven't had the bandwidth to integrate within Transformers, but I've updated the PR with a rough list of what's been done and what's left to do. If you have time to work on it by branching off my fork/branch, that would be much appreciated and I'd happy to give pointers 🙏 Otherwise, I'll be coming back to it when time frees up. Current status is still that the CTC v2 looks consistent, but the LLM variant still needs work.
